netket/operator/_local_operator/helpers.py

```diff
@@ -120,7 +120,7 @@ def canonicalize_input(
 
     # If we asked for a specific dtype, enforce it.
     if dtype is None:
-        dtype = np.promote_types(np.float32, _dtype(constant))
+        dtype = np.promote_types(float, _dtype(constant))
```

Comment:
why? 
IIRC my idea was to use `np.float32` because it's the weakest possible type, and it will not cause up-promotions of types. 

I also think this will throw some warnings if we disable x64, because this will result in double-precision instead of single precision...

Mainly, I don't want warnings to be thrown around.

Maybe we should default to double precision if x64 is enabled, and to single precision otherwise.

Reply:
Previously I noticed that we need to do change float32 to float because of [this line](https://github.com/netket/netket/blob/bb86b0234d2941f9cb4e8cf6d6afc939c8d722f1/netket/operator/_pauli_strings/base.py#L106), otherwise the dtype of `PauliStrings` defaults to float32 even if x64 is enabled.

But you made a good catch that in [this line](https://github.com/netket/netket/blob/cc5745244ba259f07c3ab0584bc57f9bb4e3dbad/netket/operator/_local_operator/helpers.py#L123) it's `promote_types` from np, not jnp. Now I'm consistently using `jnp.promote_types` in all operators, which handles weak types much saner. Also, I explicitly added `dtype = jax.dtypes.canonicalize_dtype(dtype)` to cast it to x32 when x64 is disabled. Not that even if x64 is disabled, `jnp.promote_types` can still output x64 dtypes.

Now there is no warning about x64 in the tests for operators when x64 is disabled.